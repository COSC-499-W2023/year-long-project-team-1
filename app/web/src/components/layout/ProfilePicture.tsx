/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
"use client";

import { Avatar } from "@patternfly/react-core";
import avatarImg from "@assets/pf_avatar.svg";
import { User } from "next-auth";
import Link from "next/link";
import { CSS, Stylesheet } from "@lib/utils";

const styles: Stylesheet = {
  avatar: {
    cursor: "pointer",
  },
  link: {
    display: "flex",
    flexDirection: "column",
    alignItems: "center",
    color: "var(--pf-global--Color--200)",
    margin: "0.5rem 0",
    textDecoration: "none",
  },
  username: {
    fontSize: "0.75rem",
    fontWeight: "bold",
  },
  role: {
    fontSize: "0.66rem",
    fontWeight: "lighter",
    textTransform: "capitalize",
  },
};

interface ProfilePictureProps {
  user: User;
  tooltip?: string;
  className?: string;
  style?: CSS;
  width?: string;
  link?: string;
}

export default function ProfilePicture({
  user,
  tooltip,
  className,
  style,
  width,
  link,
}: ProfilePictureProps) {
  // link to dashboard in place of user profile
  const avatar = (
    <Link
      href={link ?? "/uprofile"}
      style={{ ...styles.link, ...style }}
      title={tooltip}
    >
      <Avatar
        src={user.image || avatarImg.src}
        alt="Profile picture"
        size="md"
        className={className}
        style={{ ...styles.avatar, width, height: width }}
      />
      <span style={styles.username}>{user.username}</span>
      <span style={styles.role}>{user.role}</span>
    </Link>
  );

  return avatar;
}
