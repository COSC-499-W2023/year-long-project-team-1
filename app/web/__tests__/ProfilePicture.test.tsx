/*
 * Copyright [2023] [Privacypal Authors]
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import ProfilePicture from "@components/layout/ProfilePicture";
import { User } from "next-auth";
import { Role } from "@prisma/client";

describe("ProfilePicture", () => {
  const testUser: User = {
    id: 1,
    firstName: "Test",
    lastName: "User",
    username: "testuser",
    role: Role.CLIENT,
    email: "testuser@example.com",
    phone_number: "+1234567890",
  };

  it("matches snapshot", () => {
    const wrapper = render(<ProfilePicture user={testUser} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("renders the user's avatar", () => {
    render(<ProfilePicture user={testUser} />);
    const avatar = screen.getByAltText("Profile picture");
    expect(avatar).toBeInTheDocument();
  });

  it("renders the user's username", () => {
    render(<ProfilePicture user={testUser} />);
    const username = screen.getByText(testUser.username);
    expect(username).toBeInTheDocument();
  });

  it("renders the user's role", () => {
    render(<ProfilePicture user={testUser} />);
    const role = screen.getByText(testUser.role);
    expect(role).toBeInTheDocument();
  });

  it("has a title attribute", () => {
    render(<ProfilePicture user={testUser} tooltip="Test tooltip" />);
    const avatar = screen.getByRole("link");
    expect(avatar).toHaveAttribute("title", "Test tooltip");
  });

  it("links to the user dashboard", () => {
    render(<ProfilePicture user={testUser} />);
    const link = screen.getByRole("link");
    expect(link).toHaveAttribute("href", "/user/dashboard");
  });
});
