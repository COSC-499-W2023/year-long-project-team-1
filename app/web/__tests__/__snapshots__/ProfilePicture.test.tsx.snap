// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProfilePicture matches snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <a
        href="/user/dashboard"
        style="display: flex; flex-direction: column; align-items: center; margin: 0.5rem 0px; text-decoration: none;"
      >
        <img
          alt="Profile picture"
          class="pf-v5-c-avatar pf-m-md"
          src="/img.jpg"
          style="cursor: pointer;"
        />
        <span
          style="font-size: 0.75rem; font-weight: bold;"
        >
          testuser
        </span>
        <span
          style="font-size: 0.66rem; font-weight: lighter; text-transform: capitalize;"
        >
          CLIENT
        </span>
      </a>
    </div>
  </body>,
  "container": <div>
    <a
      href="/user/dashboard"
      style="display: flex; flex-direction: column; align-items: center; margin: 0.5rem 0px; text-decoration: none;"
    >
      <img
        alt="Profile picture"
        class="pf-v5-c-avatar pf-m-md"
        src="/img.jpg"
        style="cursor: pointer;"
      />
      <span
        style="font-size: 0.75rem; font-weight: bold;"
      >
        testuser
      </span>
      <span
        style="font-size: 0.66rem; font-weight: lighter; text-transform: capitalize;"
      >
        CLIENT
      </span>
    </a>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
